src/test/scala/sigmastate/utxo/examples/FsmExampleSpecification.scala

```diff
@@ -1,20 +1,254 @@
 package sigmastate.utxo.examples
 
-import sigmastate.helpers.SigmaTestingCommons
-
-/**
-  * Similarly to MAST example, we can do more complex contracts, e.g. ones with cycles. For example, we can do a
-  * contract described as a finite state machine.
-  * Let's consider that a input in a finite state machine is its script, or its hash. Then assume a machine is
-  * described as a table of transitions, an example is below:
-  * state1 | script_hash1 | state2
-  * state2 | script_hash2 | state3
-  * state2 | script_hash3 | state1
-  * state3 | script_hash4 | F
-  *
-  * where F is a final state, which denotes end of a contract execution
-  */
+import org.ergoplatform.ErgoBox.{R3, R4}
+import org.ergoplatform._
+import scorex.crypto.authds.{ADKey, ADValue}
+import scorex.crypto.authds.avltree.batch.{BatchAVLProver, Insert, Lookup}
+import scorex.crypto.hash
+import scorex.crypto.hash.{Blake2b256, Digest32}
+import sigmastate.Values._
+import sigmastate._
+import sigmastate.helpers.{ErgoLikeProvingInterpreter, SigmaTestingCommons}
+import sigmastate.serialization.ValueSerializer
+import sigmastate.utxo._
+
 
 class FsmExampleSpecification extends SigmaTestingCommons {
 
+  /**
+    * Similarly to the MAST-like example (in the MASTExampleSpecification class), we can do more complex contracts,
+    * e.g. ones with cycles. For example, we can do a contract described as a finite state machine.
+    *
+    * Let's consider that a input in a finite state machine is a script associated with a transition (so the transition
+    * could happen only if the script is satisfied by a spending proof in a spending transaction), or a hash of the
+    * script. Then we assume that a machine is described via a table of transitions, for example, as below:
+    *
+    * state1 | script_hash1 | state2
+    * state2 | script_hash3 | state3
+    * state2 | script_hash2 | state1
+    * state3 | script_hash4 | -
+    *
+    * where state3 is a final state, which denotes end of a contract execution (and by executing a script with the hash
+    * script_hash4 it is possible to create any box as a result).
+    *
+    * We put the table description into an authenticated data structure (such as the Merkle tree, we're using
+    * authenticated AVL+ trees though). We require root digest of the tree to be preserved. We require FSM framework
+    * script to be preserved as well. So to trigger a transition from one state a coin remembers to another one,
+    * a spending transaction should create a coin with preserved root digest, and some (valid) new state, show
+    * a proof for a corresponding transition description in the tree, and show a proof for transition script (e.g. a
```

Comment:
@kushti why not use Box based terminology. It is more consistent with the language. It also sounds better in the sentences above where. It seems natural for boxes to remember state.

Reply:
What can be done with the box in final state? Which is final state: state3 or - ?